ComponentKit/Core/Scope/CKDetectComponentScopeCollisions.mm

```diff
@@ -0,0 +1,41 @@
+/*
+ *  Copyright (c) 2014-present, Facebook, Inc.
+ *  All rights reserved.
+ *
+ *  This source code is licensed under the BSD-style license found in the
+ *  LICENSE file in the root directory of this source tree. An additional grant
+ *  of patent rights can be found in the PATENTS file in the same directory.
+ *
+ */
+
+#import "CKDetectComponentScopeCollisions.h"
+
+#import <queue>
+
+#import "CKAssert.h"
+#import "CKComponentInternal.h"
+
+void CKDetectComponentScopeCollisions(const CKComponentLayout &layout)
+{
+#if DEBUG
+  std::queue<const CKComponentLayout> queue;
+  NSMutableSet<NSNumber *> *previouslySeenScopeFrameTokens = [NSMutableSet new];
+  queue.push(layout);
+  while (!queue.empty()) {
+    auto currentLayout = queue.front();
+    queue.pop();
+    NSNumber *scopeFrameToken = [currentLayout.component scopeFrameToken];
+    if (scopeFrameToken && [previouslySeenScopeFrameTokens containsObject:scopeFrameToken]) {
+      CKCFailAssert(@"Scope collision. Attempting to create duplicate scope for component %@", [currentLayout.component class]);
```

Comment:
An improvement here would be to include a "backtrace" of components leading to the collision.